src/VisualStudio/Core/Def/Implementation/Experimentation/KeybindingResetDetector.cs

```diff
@@ -217,37 +261,87 @@ private void ShowGoldBar()
+
+                var hr_suspend = await QueryStatusAsync(cmds_suspend).ConfigureAwait(false);
+                var hr_resume = await QueryStatusAsync(cmds_resume).ConfigureAwait(false);
+
                 // In the case of an error when attempting to get the status, pretend that ReSharper isn't enabled. We also
                 // shut down monitoring so we don't keep hitting this.
-                FatalError.ReportWithoutCrash(Marshal.GetExceptionForHR(hr));
-                Shutdown();
-                return ReSharperStatus.NotInstalledOrDisabled;
+                if (ErrorHandler.Failed(hr_suspend))
+                {
+                    FatalError.ReportWithoutCrash(Marshal.GetExceptionForHR(hr_suspend));
+                    await ShutdownAsync().ConfigureAwait(false);
+
+                    return ReSharperStatus.NotInstalledOrDisabled;
+                }
+
+                else if (ErrorHandler.Failed((hr_resume)))
+                {
+                    FatalError.ReportWithoutCrash(Marshal.GetExceptionForHR(hr_resume));
+                    await ShutdownAsync().ConfigureAwait(false);
+
+                    return ReSharperStatus.NotInstalledOrDisabled;
+                }
+
+                // When ReSharper is suspended, the ReSharper_Resume command is Enabled and not Invisible
+                if (((OLECMDF)cmds_resume[0].cmdf).HasFlag(OLECMDF.OLECMDF_ENABLED) && !((OLECMDF)cmds_resume[0].cmdf).HasFlag(OLECMDF.OLECMDF_INVISIBLE))
+                {
+                    return ReSharperStatus.Suspended;
+                }
+
+                // When ReSharper is running, the ReSharper_Suspend command is Enabled and not Invisible
+                if (((OLECMDF)cmds_suspend[0].cmdf).HasFlag(OLECMDF.OLECMDF_ENABLED) && !((OLECMDF)cmds_suspend[0].cmdf).HasFlag(OLECMDF.OLECMDF_INVISIBLE))
+                {
+                    return ReSharperStatus.Enabled;
+                }
+
+                //otherwise sleep for a bit and check again
+                await Task.Delay(TimeSpan.FromSeconds(2), cancellationToken).ConfigureAwait(false);
             }
 
-            // When ReSharper is enabled, the ReSharper_Suspend command has the Enabled | Supported flags. When disabled, it has Invisible | Supported.
-            return ((OLECMDF)cmds[0].cmdf).HasFlag(OLECMDF.OLECMDF_ENABLED) ? ReSharperStatus.Enabled : ReSharperStatus.Suspended;
+            async Task<int> QueryStatusAsync(OLECMD[] cmds)
+            {
+                await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+                return _oleCommandTarget.QueryStatus(ReSharperCommandGroup, (uint)cmds.Length, cmds, IntPtr.Zero);
+            }
+
+            async Task EnsureOleCommandTargetAsync()
+            {
+                if (_oleCommandTarget != null)
+                {
+                    return;
+                }
+
+                await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```

Comment:
❕ Need to check the `CancellationToken` after a switch to the main thread.

```suggestion
                await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
                cancellationToken.ThrowIfCancellationRequested();
```